### PR TITLE
Feature/52636 search facets tree view

### PR DESCRIPTION
**Pull request by InnaBoitsun**
Ipmplement Category Facets in Tree view for Search Page and Category Browse pages
Mobile View for Search and Category pages
TODO: Query page
i5q5Ev97fN

[AB#52636](https://dev.azure.com/orckestra001/da965986-c85b-43be-9915-219568b000e5/_workitems/edit/52636)
[AB#52650](https://dev.azure.com/orckestra001/da965986-c85b-43be-9915-219568b000e5/_workitems/edit/52650)
[AB#52791](https://dev.azure.com/orckestra001/da965986-c85b-43be-9915-219568b000e5/_workitems/edit/52791)

### CODE REVIEW

File: src/Orckestra.Composer.Search/Orckestra.Composer.Search.csproj [hunk 1]
```diff
@@ -184,7 +186,7 @@
     </None>
   </ItemGroup>
   <ItemGroup>
-    <Folder Include="Extensions\" />
+    <Folder Include="Extentions\" />
```
Comment (napernik) [inline src/Orckestra.Composer.Search/Orckestra.Composer.Search.csproj:189]: *Extensions

File: src/Orckestra.Composer.Search/Repositories/SearchRepository.cs [hunk 2]
```diff
@@ -179,5 +179,38 @@ protected virtual QuerySorting BuildQuerySortings(SearchCriteria criteria)
...
+                request.FacetPredicates = criteria.SelectedFacets
+                        .Where(sf => !sf.Name.StartsWith(SearchConfiguration.CategoryFacetFiledNamePrefix))
+                        .Select(FacetPredicateFactory.CreateFacetPredicate)
+                        .Where(fp => fp != null).ToList();
+            }
+            request.InventoryLocationIds = criteria.InventoryLocationIds;
+            request.AutoCorrect = criteria.AutoCorrect;
+            request.AvailabilityDate = criteria.AvailabilityDate;
+
+            var results = await ExecuteProductSearchRequestAsync(request).ConfigureAwait(false);
```
Comment (napernik) [inline src/Orckestra.Composer.Search/Repositories/SearchRepository.cs:211]: When there's only one async call in the method, and it is the latest one, you can remove the "async" keyword and return the subtast task itself:

```
 public virtual Task<ProductSearchResult> GetCategoryFacetCountsAsync(SearchCriteria criteria)
{
    ....
    return ExecuteProductSearchRequestAsync(request);
}
```

File: src/Orckestra.Composer.Search/SearchConfiguration.cs [hunk 3]
```diff
@@ -194,5 +194,10 @@ public static ProviderRegistry<ISelectedFacetProvider> SelectedFacetProviderRegi
         /// Whether search term should be automatically corrected
         /// </summary>
         public static bool AutoCorrectSearchTerms { get; set; } = true;
+
+        /// <summary>
+        /// The prefix for the the category facet field name to indentify what facets are for categories
```
Comment (napernik) [inline src/Orckestra.Composer.Search/SearchConfiguration.cs:199]: *identify which

File: src/Orckestra.Composer.Search/Services/BaseSearchViewService.cs [hunk 4]
```diff
@@ -209,6 +214,44 @@ protected virtual async Task<ProductSearchResultsViewModel> CreateProductSearchR
...
+        }
+
+        protected virtual CategoryFacetValuesTree BuildCategoryFacetValuesTree(IList<Facet> facets, 
+            SelectedFacets selectedFacets, 
+            CategoryFacetCounts categoryCounts)
+        {
+            var categories = CategoryRepository.GetCategoriesTreeAsync(new GetCategoriesParam
+            {
+                Scope = ComposerContext.Scope
+            }).ConfigureAwait(false).GetAwaiter().GetResult();
```
Comment (napernik) [inline src/Orckestra.Composer.Search/Services/BaseSearchViewService.cs:250]: this method can be made async instead

File: src/Orckestra.Composer.Search/Services/BaseSearchViewService.cs [hunk 5]
```diff
@@ -540,6 +584,12 @@ protected virtual async Task<ProductSearchResultsViewModel> SearchAsync(TParam p
                 SearchResult = searchResult
             };
 
+            if (param.Criteria.SelectedFacets != null && 
+                param.Criteria.SelectedFacets.Any(s => !string.IsNullOrWhiteSpace(s.Name) && s.Name.StartsWith(SearchConfiguration.CategoryFacetFiledNamePrefix)))
```
Comment (napernik) [inline src/Orckestra.Composer.Search/Services/BaseSearchViewService.cs:588]: Can be simplified

s => !string.IsNullOrWhiteSpace(s.Name) && s.Name.StartsWith(SearchConfiguration.CategoryFacetFiledNamePrefix)

s => s.Name?.StartsWith(SearchConfiguration.CategoryFacetFiledNamePrefix) ?? false

File: src/Orckestra.Composer.Search/Services/CategoryBrowsingViewService.cs [hunk 6]
```diff
@@ -85,13 +83,45 @@ public virtual async Task<CategoryBrowsingViewModel> GetCategoryBrowsingViewMode
...
+                }
+            }
+
             return viewModel;
         }
 
+        private void BuildCategoryUrlsForTreeItem(GetCategoryBrowsingViewModelParam param, CategoryFacetValuesTreeItem item)
+        {
+            item.CategoryUrl = GetCategoryUrl(item.CategoryId, param);
+            if (item.Items != null)
```
Comment (napernik) [inline src/Orckestra.Composer.Search/Services/CategoryBrowsingViewService.cs:107]: Can be simplified:
```
item.Items?.ForEach(subItem => BuildCategoryUrlsForTreeItem(param, subItem));
```

**Comment by InnaBoitsun** [comment]
@napernik could you please do final review. Thanks

File: src/Orckestra.Composer.Search/Services/CategoryBrowsingViewService.cs [hunk 7]
```diff
@@ -73,15 +71,72 @@ public virtual async Task<CategoryBrowsingViewModel> GetCategoryBrowsingViewMode
...
 
+        private CategoryFacetValuesTreeNode CategoryRootNode(List<CategoryFacetValuesTreeNode> nodes, string categoryId)
+        {
+            if (nodes == null) return null;
+            var rootCategory = nodes.FirstOrDefault(c => c.CategoryId == categoryId);
+            if (rootCategory == null)
+            {
+                foreach (var childNode in nodes)
+                {
+                    rootCategory = CategoryRootNode(childNode.ChildNodes, categoryId);
```
Comment (neexite) [inline src/Orckestra.Composer.Search/Services/CategoryBrowsingViewService.cs:111]: Smth wrong, by code it looks like root category assigned from latest  childNode, and all previous calculation useless

File: src/Orckestra.Composer.Website/App_Code/FacetHelpers.cshtml [hunk 8]
```diff
@@ -119,55 +110,104 @@
...
+    <div v-if="facet.FacetType == '@(FacetType.SingleSelect)'">
+        @SingleFacet()
+    </div>
+    <div v-if="facet.FacetType == '@(FacetType.MultiSelect)'">
+        @MultiSelectFacet()
+    </div>
+    <div v-if="facet.FacetType == '@(FacetType.Range)'">
+        @RangeFacet()
+    </div>
+    <div v-if="facet.FacetType != '@(FacetType.Range)'">
```
Comment (neexite) [inline src/Orckestra.Composer.Website/App_Code/FacetHelpers.cshtml:124]: Is it normal ShowMoreLessFacets shoed when SingleSelect or MultiSelect?

File: src/Orckestra.Composer.Website/UI.Package/LocalizedStrings/List-Search.resx [hunk 9]
```diff
@@ -300,4 +300,13 @@
   <data name="T_SearchResultsFoundIn" xml:space="preserve">
     <value>{0} Found ({1})</value>
   </data>
+  <data name="B_ApplyFilters" xml:space="preserve">
```
Comment (neexite) [inline src/Orckestra.Composer.Website/UI.Package/LocalizedStrings/List-Search.resx:303]: Do we have theses texts fro FR?

File: src/Orckestra.Composer.Website/UI.Package/Sass/Custom/Search/_facets.scss [hunk 10]
```diff
@@ -32,3 +107,99 @@
...
+            .close {
+                padding: 20px;
+            }
+        }
+
+        .facets-modal-body {
+            position: relative;
+            overflow-y: auto;
+            flex: 1 1 auto;
+           //TODO min-height: subtract(100vh, 120px);
```
Comment (neexite) [inline src/Orckestra.Composer.Website/UI.Package/Sass/Custom/Search/_facets.scss:170]: Is it finished?(TODO:)

**Review by neexite — CHANGES_REQUESTED** [review]

File: src/Orckestra.Composer.Website/App_Code/FacetHelpers.cshtml (hunk 8 above)
Comment (InnaBoitsun) [reply src/Orckestra.Composer.Website/App_Code/FacetHelpers.cshtml:124]: yes, for Range Facet we do not show this

File: src/Orckestra.Composer.Website/UI.Package/Sass/Custom/Search/_facets.scss (hunk 10 above)
Comment (InnaBoitsun) [reply src/Orckestra.Composer.Website/UI.Package/Sass/Custom/Search/_facets.scss]: no, but I added more information into TODO to invetigate later. It doesn't work just on my iPhone xS, but looks like works on others phones.

File: src/Orckestra.Composer.Search/Services/CategoryBrowsingViewService.cs (hunk 7 above)
Comment (InnaBoitsun) [reply src/Orckestra.Composer.Search/Services/CategoryBrowsingViewService.cs:111]: This is specific case for Category Page. By default we have the whole tree for Categories. 
In CategoryBrowsingViewService we rewrite the root of tree and set it to the Category page Root.
First client wanted to render all tree from 1 level root for category page. Then they changed mind to have tree started from Category page category.
Other calculations for Category Urls we do for the new rewrited Tree.

**Review by neexite — APPROVED** [review]

**Pull request merged**